buffer/src/main/java/io/netty/buffer/PoolChunk.java

```diff
@@ -182,23 +214,58 @@
         this.arena = arena;
         this.memory = memory;
         this.offset = offset;
-        memoryMap = null;
-        depthMap = null;
-        subpages = null;
-        subpageOverflowMask = 0;
         pageSize = 0;
         pageShifts = 0;
-        maxOrder = 0;
-        unusable = (byte) (maxOrder + 1);
+        runsAvailMap = null;
+        runsAvail = null;
+        subpages = null;
         chunkSize = size;
-        log2ChunkSize = log2(chunkSize);
-        maxSubpageAllocs = 0;
         cachedNioBuffers = null;
     }
 
     @SuppressWarnings("unchecked")
-    private PoolSubpage<T>[] newSubpageArray(int size) {
-        return new PoolSubpage[size];
+    private static PriorityQueue<Long>[] newRunsAvailqueueArray(int size) {
+        PriorityQueue<Long>[] queueArray = new PriorityQueue[size];
+        for (int i = 0; i < queueArray.length; i++) {
+            queueArray[i] = new PriorityQueue<Long>();
+        }
+        return queueArray;
+    }
+
+    private void insertAvailRun(int runOffset, int pages, Long handle) {
+        int pageIdxFloor = arena.pages2pageIdxFloor(pages);
+        PriorityQueue<Long> queue = runsAvail[pageIdxFloor];
+        queue.offer(handle);
+
+        //insert first page of run
+        runsAvailMap.put(runOffset, handle);
+        if (pages > 1) {
+            //insert last page of run
+            runsAvailMap.put(runOffset + pages - 1, handle);
+        }
+    }
+
+    private void removeAvailRun(long handle) {
+        int pageIdxFloor = arena.pages2pageIdxFloor(runPages(handle));
+        PriorityQueue<Long> queue = runsAvail[pageIdxFloor];
+        removeAvailRun(queue, handle);
+    }
+
+    private void removeAvailRun(PriorityQueue<Long> queue, long handle) {
+        queue.remove(handle);
+
+        int runOffset = runOffset(handle);
+        int pages = runPages(handle);
+        //remove first page of run
+        runsAvailMap.remove(runOffset);
+        if (pages > 1) {
+            //remove last page of run
+            runsAvailMap.remove(runOffset + pages - 1);
```

Comment:
nit: consider adding `lastPage()` meted and use it here and above as the calculation is the same